adjoint/transpose transcend material concerns

Sacha0 opened the pull request:
This pull request is the next step in the JuliaLang/LinearAlgebra.jl#57 series after JuliaLang/julia#24969, JuliaLang/julia#25083, JuliaLang/julia#25125, JuliaLang/julia#25148, and JuliaLang/julia#25217. Specifically, this pull request makes `adjoint`/`transpose` uniformly lazy and extends ~`collect`~ `copy` for materialization.

Remaining tasks, for this or a downstream pull request:
1. ~Transform most `Adjoint`/`Transpose` calls to `adjoint`/`transpose`.~ done here
3. ~Lower `'` directly to `adjoint` (thereby fixing method overwrites during build).~ done here
2. Make `Adjoint`/`Transpose` behave like proper constructors.

Let's see what CI says. Thanks and best!

Sacha0 committed:
Make adjoint/transpose lazy with copy for materialization.

mbauman commented on base/linalg/bitarray.jl:
```diff
@@ -123,7 +123,7 @@ end
 ## Structure query functions
 
-issymmetric(A::BitMatrix) = size(A, 1)==size(A, 2) && count(!iszero, A - transpose(A))==0
+issymmetric(A::BitMatrix) = size(A, 1)==size(A, 2) && count(!iszero, A - collect(A'))==0
```
Just curious: is your rationale for `collect`ing here to ensure we still hit an optimized `-(::BitMatrix, ::BitMatrix)` method?  If that's the case it may be worth keeping note of this so we can revisit this PR for more `::Adjoint` method specializations in the future.

Sacha0 replied:
Partly, yes :). During the relevant transformation pass, preserving behavior was my primary concern. And for some array types, the rewrite above was strictly necessary to preserve behavior. (For example, `S + Adjoint(S)` for sparse matrix `S` hits an `AbstractArray` `+` fallback yielding an `Array` rather than the former `S + adjoint(S)`'s `SparseMatrixCSC`. Hence the transformation `S + adjoint(S)` -> `S + collect(S')` rather than `S + S'`, though `S + S'` should work eventually.) And indeed, after the gross functionality is complete quite a bit of cleanup will be in order :). https://github.com/JuliaLang/julia/issues/25331 is my present breadcrumb for this particular issue.

mbauman commented on base/linalg/generic.jl:
```diff
@@ -986,7 +986,7 @@ function ishermitian(A::AbstractMatrix)
     end
     for i = indsn, j = i:last(indsn)
-        if A[i,j] != adjoint(A[j,i])
+        if A[i,j] != Adjoint(A[j,i])
```
Why are you favoring the uppercase constructors in these cases?

Sacha0 replied:
The simple answer is that this pull request presently represents an intermediate state which will change substantially on next push. Specifically, most uppercase calls should then disappear :).

mbauman commented on base/number.jl:
```diff
@@ -67,9 +67,12 @@ function getindex(x::Number, I::Integer...)
 last(x::Number) = x
-copy(x::Number) = x  # some code treats numbers as collection-like
+copy(x::Number) = x
+collect(x::Number) = x
```
This may be worth calling out and thinking through a bit more.  I expected `collect(1)` to return `[1]` since numbers are iterable "collections". In fact, right now `collect(1) = fill(1)` (that is, a zero-d array).

 I get why this change is needed, but I'm not entirely sold that it's the right answer.

Sacha0 replied:
Agreed, I'm not entirely sold on it either :).

JeffBezanson replied:
I'm also -1 on this definition.

Sacha0 committed:
Replace Adjoint/Transpose with adjoint/transpose throughout test/.

Sacha0 committed:
Replace Adjoint/Transpose with adjoint/transpose throughout base/.

Sacha0 commented:
(Marking triage for `collect` discussion.)

Sacha0 committed:
Replace Adjoint/Transpose with adjoint/transpose throughout deprecations.

vtjnash commented:
From discussion on triage, there was some consensus to go with the following:
- deprecate `collect` name:
  - use `Array` constructor instead in some cases (usually for realizing lazy operations like iterators)
  - use `copy` in other cases (probably most of these)
    - define `copy(x::AbstractArray) = Array(x)` (or some equivalent like `copyto!(similar(Array, x), x)`)
    - encourage specializations of `copy` when there's a better definition which can fold away extra wrappers or, failing that, can return an object of the same type

Sacha0 committed:
Replace Adjoint/Transpose with adjoint/transpose in IterativeEigensolvers.

Sacha0 committed:
Replace Adjoint/Transpose with adjoint/transpose in SuiteSparse.

Sacha0 commented:
Rebased and completed rewrites to `adjoint`/`transpose` in base, test, and stdlib. Passing sparse, linalg, and stdlib test suites locally. So this pull request should be in shape after a sweep to replace `collect` methods/calls with `copy` as per triage. Best!

Sacha0 commented:
Rebased, squashed out all intermediate commits in the adjoint/transpose lazification process, and added a commit transforming collect to copy as per triage. The last commit is #25418, which I will drop once #25418 merges. Passed all impacted test suites locally. Best!

Sacha0 committed:
Lower ' directly to adjoint.

Sacha0 commented:
Added a commit lowering `'` directly to `adjoint` and further cleaned up the commit history; the lowering change incidentally fixes the method overwrite warning during build mentioned in https://github.com/JuliaLang/julia/pull/25212. I plan to make the `Adjoint`/`Transpose` constructor changes in a separate pull request. So this pull request should now be in shape. Best!

mbauman commented on base/pkg/resolve/maxsum.jl:
```diff
@@ -122,7 +122,7 @@ mutable struct Graph
 
                         bm = trues(spp[p1], spp[p0])
-                        bmt = adjoint(bm)
+                        bmt = trues(spp[p0], spp[p1])
```
Wow, your re-write here is _way_ better.  Bravo for not just zoning out (like I did when this change caught me off guard).

Sacha0 replied:
One small step towards atonement for the horde of ghastly transformations I introduced in #24969 😉.

mbauman reviewed (approved):
LGTM!  Very nice work.

Sacha0 commented:
Rebased out the last commit with #25418 in. With Matt and CI approving, absent objections or requests for time I plan to merge these changes Sunday evening PT or later. Best! :)

Sacha0 commented:
Rebased out conflict. Absent objections I plan to merge these changes once CI clears. Best!

Sacha0 commented:
(AV and FreeBSD failures are timeouts without apparent issue prior.)

Sacha0 commented:
Thanks all! :)
